internal/directives/file_deleter.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+func init() {
+	builtins.RegisterPromotionStepRunner(newFileDeleter(), nil)
+}
+
+type fileDeleter struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+func newFileDeleter() PromotionStepRunner {
+	r := &fileDeleter{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+func (f *fileDeleter) Name() string {
+	return "delete"
+}
+
+func (f *fileDeleter) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	// Validate the configuration against the JSON Schema.
+	if err := validate(f.schemaLoader, gojsonschema.NewGoLoader(stepCtx.Config), f.Name()); err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}, err
+	}
+
+	// Convert the configuration into a typed object.
+	cfg, err := ConfigToStruct[DeleteConfig](stepCtx.Config)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not convert config into %s config: %w", f.Name(), err)
+	}
+
+	return f.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+func (f *fileDeleter) runPromotionStep(
+	_ context.Context,
+	stepCtx *PromotionStepContext,
+	cfg DeleteConfig,
+) (PromotionStepResult, error) {
+	// join with a regular join
+	absPath, err := checkAbsPath(stepCtx.WorkDir, cfg.Path)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not secure join path %q: %w", cfg.Path, err)
+	}
+
+	if checkSymlink(absPath) {
+		err := os.Remove(absPath)
+		if err != nil {
+			return PromotionStepResult{}, err
+		}
+	} else {
+		pathToDelete, err := securejoin.SecureJoin(stepCtx.WorkDir, cfg.Path)
+		if err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("could not secure join path %q: %w", cfg.Path, err)
+		}
+
+		if err = removePath(pathToDelete); err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("failed to delete %q: %w", cfg.Path, sanitizePathError(err, stepCtx.WorkDir))
+		}
+	}
+
+	return PromotionStepResult{Status: kargoapi.PromotionPhaseSucceeded}, nil
+}
+
+func checkSymlink(path string) bool {
+	fi, err := os.Lstat(path)
+	if err != nil {
+		return false
+	}
+
+	return fi.Mode()&os.ModeSymlink != 0
+}
+
+func checkAbsPath(workDir string, path string) (string, error) {
+	absBase, err := filepath.Abs(workDir)
+	if err != nil {
+		return "", err
+	}
+
+	fullPath := filepath.Join(workDir, path)
+	absPath, err := filepath.Abs(fullPath)
+	if err != nil {
+		return "", err
+	}
+
+	// Get the relative path from base to the requested path
+	// If the requested path tries to escape, this will return
+	// an error or a path starting with "../"
+	relPath, err := filepath.Rel(absBase, absPath)
+	if err != nil {
+		return "", err
+	}
+
+	// Check if path attempts to escape
+	if strings.HasPrefix(relPath, "..") {
```

Comment:
This isn't the right way to check for this.

`sub-dir/../../something-i-should-not-access` accomplishes the same thing as `../something-i-should-not-access`, but the former isn't caught by this logic.

Reply:
Nevermind... seeing where `relPath` comes from, containing "..", but not starting with it is an impossibility.